### PR TITLE
[SPARK-41489][SQL] Assign name to _LEGACY_ERROR_TEMP_2415

### PR DESCRIPTION
**Pull request by itholic**
### What changes were proposed in this pull request?

This PR proposes to assign name to _LEGACY_ERROR_TEMP_2415, "DATATYPE_MISMATCH.FILTER_NOT_BOOLEAN".

### Why are the changes needed?

We should assign proper name to _LEGACY_ERROR_TEMP_*

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

`./build/sbt "sql/testOnly org.apache.spark.sql.SQLQueryTestSuite*`

### CODE REVIEW

File: core/src/main/resources/error/error-classes.json [hunk 1]
```diff
@@ -933,6 +933,12 @@
     ],
     "sqlState" : "42604"
   },
+  "INVALID_TYPE_FOR_FILTER_EXPR" : {
+    "message" : [
+      "Filter expression '<filter>' of type <type> is not a boolean."
+    ],
+    "sqlState" : "428H2"
```
Comment (itholic) [inline core/src/main/resources/error/error-classes.json:940]: FYI: referred to https://github.com/apache/spark/tree/master/core/src/main/resources/error

**Comment by itholic** [comment]
cc @MaxGekk @srielau @cloud-fan

File: core/src/main/resources/error/error-classes.json [hunk 2]
```diff
@@ -933,6 +933,12 @@
     ],
     "sqlState" : "42604"
   },
+  "INVALID_TYPE_FOR_FILTER_EXPR" : {
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:936]: Is it a specific case of `DATATYPE_MISMATCH.UNEXPECTED_INPUT_TYPE`. Can't you re-use it here?

Comment (srielau) [reply core/src/main/resources/error/error-classes.json:936]: It isn't really a parameter. It's debatable whether it should be a subclass. But DATATYPE_MISMATCH is really turning into a catchall. I would either leave it as is, or make it:
DATATYPE_MISMATCH.FILTER_NOT_BOOLEAN or so..

Comment (itholic) [reply core/src/main/resources/error/error-classes.json]: Yes, I agree that it's not really related to an input parameter.

Let me put it in the `DATATYPE_MISMATCH` and rename it as `FILTER_NOT_BOOLEAN` for now.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 3]
```diff
@@ -355,8 +355,9 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
             }
           case f: Filter if f.condition.dataType != BooleanType =>
             f.failAnalysis(
-              errorClass = "_LEGACY_ERROR_TEMP_2415",
+              errorClass = "DATATYPE_MISMATCH.FILTER_NOT_BOOLEAN",
               messageParameters = Map(
+                "sqlExpr" -> f.expressions.map(toSQLExpr).mkString(","),
                 "filter" -> f.condition.sql,
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:361]: Please, use `toSQLExpr` here

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala [hunk 4]
```diff
@@ -355,8 +355,9 @@ trait CheckAnalysis extends PredicateHelper with LookupCatalog with QueryErrorsB
...
                 "type" -> f.condition.dataType.catalogString))
```
Comment (MaxGekk) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala:362]: Use `toSQLType()`

File: core/src/main/resources/error/error-classes.json [hunk 5]
```diff
@@ -265,6 +265,11 @@
           "Input to <functionName> should all be the same type, but it's <dataType>."
         ]
       },
+      "FILTER_NOT_BOOLEAN" : {
+        "message" : [
+          "Filter expression '<filter>' of type <type> is not a boolean."
```
Comment (MaxGekk) [inline core/src/main/resources/error/error-classes.json:270]: Could you remove '' around <filter>

**Review by MaxGekk — CHANGES_REQUESTED** [review]

**Comment by itholic** [comment]
Thanks @MaxGekk . Resolved the comments

**Review by MaxGekk — APPROVED** [review]

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @itholic and @srielau for review.

**Comment by MaxGekk** [comment]
@itholic Could you backport this to 3.4, please.

**Comment by itholic** [comment]
Backport: https://github.com/apache/spark/pull/39829

**Pull request closed without merging**